File: Backend/src/main/java/lk/fuel_app/service/CustomerService.java
package lk.fuel_app.service;

import lk.fuel_app.dto.FuelAvailabilityDTO;
import lk.fuel_app.entity.Customer;
import lk.fuel_app.entity.CustomerFuelStation;
import lk.fuel_app.entity.FuelConsumption;
import lk.fuel_app.entity.Vehicle;

import java.util.List;

public interface CustomerService {
    Customer addCustomer(Customer customer);

    Customer getCustomer(String email, String contactNumber);

    boolean deleteCustomer(String id);

    Customer updateCustomer(Customer customer, String id);

    Customer getCustomerByVehicle(String vehicle);

    List<CustomerFuelStation> getPumpedAmounts(String id);

    Customer sendOTP(String email, String contactNumber);

    List<FuelAvailabilityDTO> fuelAvailability(String place, String orderBy);

    FuelConsumption addFuelConsumption(FuelConsumption fuelConsumption);

    List<FuelConsumption> getFuelConsumptions(String id);

    boolean deleteFuelConsumption(String id);

    List<FuelAvailabilityDTO> getFuelAvailabilityM(String id);

    FuelAvailabilityDTO getFuelConsumptionsM(String vehicle);

    Vehicle regenerateQR(String vehicle);
}

File: Frontend/src/app/_service/customer.service.ts
import {Injectable} from '@angular/core';
import {environment} from '../../environments/environment';
import {HttpClient, HttpClientModule} from '@angular/common/http';
import {Observable} from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class CustomerService {

  customer;

  constructor(private http: HttpClient) {
  }

  addCustomer(customer): Observable<any> {
    return this.http.post<any>(environment.backend_url + '/customer/addCustomer', customer);
  }

  updateCustomer(customer): Observable<any> {
    return this.http.put<any>(environment.backend_url + '/customer/updateCustomer/' + customer.nic, customer);
  }

  deleteCustomer(id): Observable<any> {
    return this.http.delete<any>(environment.backend_url + '/customer/deleteCustomer/' + id);
  }

  getCustomer(email, contactNumber): Observable<any> {
    return this.http.get<any>(environment.backend_url + '/customer/getCustomer/' + email + '/' + contactNumber);
  }

  getCustomerByVehicle(vehicle): Observable<any> {
    return this.http.get<any>(environment.backend_url + '/customer/getCustomerByVehicle/' + vehicle);
  }

  getPumpedAmounts(id): Observable<any> {
    return this.http.get<any>(environment.backend_url + '/customer/getPumpedAmounts/' + id);
  }

  sendOTP(email, contactNumber): Observable<any> {
    return this.http.get<any>(environment.backend_url + '/customer/sendOTP/' + email + '/' + contactNumber);
  }

  getFuelAvailability(place, orderBy): Observable<any> {
    return this.http.get<any>(environment.backend_url + "/customer/fuelAvailability/" + place + "/" + orderBy);
  }

  addFuelConsumption(fuelConsumption): Observable<any> {
    return this.http.post<any>(environment.backend_url + '/customer/addFuelConsumption', fuelConsumption);
  }

  deleteFuelConsumption(id): Observable<any> {
    return this.http.delete<any>(environment.backend_url + '/customer/deleteFuelConsumption/' + id);
  }

  getFuelConsumptions(id): Observable<any> {
    return this.http.get<any>(environment.backend_url + "/customer/getFuelConsumptions/" + id);
  }

  newCustomer() {
    return {
      nic: '',
      name: '',
      address: '',
      district: '',
      fuelStationPlace: {
        id: ''
      },
      vehicle: {
        vehicleType: {
          id: ''
        },
        chassisNumber: '',
        vehicleNumber: '',
        fuelType: '',
      },
      quota: 0,
      appUser: {
        email: '',
        contactNumber: '',
        password: ''
      }
    };
  }
}
